### DISCUSSION TITLE
Cairo+Pango front/xpm rendering on X11

### DISCUSSION THREAD

**Discussion by dannye** [fltk/fltk#1057] [Q&A] [answered]
Starting this out as a discussion rather than an issue because I'm not totally sure if this is a bug or not.

In #963 it was suggested that enabling Cairo could solve an unrelated/unimportant issue I was having, but the main reason I didn't choose this option is because enabling Cairo messed up all font rendering in my application ([example](https://github.com/fltk/fltk/issues/963#issuecomment-2084289983)).

In #1052 it was suggested yet again that enabling Cairo could fix another unrelated issue. While testing that, I also realized that enabling Cairo fixes dashed/dotted line styles, which are \*awfully\* broken on X11 without Cairo ([example](https://github.com/fltk/fltk/issues/1052#issuecomment-2311059052)).

So the reasons to use Cairo are really stacking up, but the terrible font problem remains!

I have just figured out why this is: If a user application calls `Fl::set_fonts(NULL)` (just for the sake of querying the number of system fonts) a second time after already configuring a custom font via `Fl::set_font(font, name)`, then font rendering will be "broken", and this issue only happens on X11+Cairo.

"Redundantly" calling `Fl::set_fonts(NULL)` doesn't cause this same issue on Windows, Mac, or X11 without Cairo.

This is pretty simple to avoid in my application. I just call `Fl::set_fonts(NULL)` once and cache the result.

But I was under the impression that `Fl::set_fonts(NULL)` is supposed to be idempotent, and if that's true then I think this is an FLTK bug. Any insight is appreciated.

---

Not related to fonts, the other issue I have with enabling Cairo on X11 is that it creates a "smearing" effect on any xpm icon which uses its right edge and bottom edge (ie, does not have a 1-px transparent margin on the right and bottom edges).

This is what it looks like on X11 without Cairo (scaled up for detail):

![image](https://github.com/user-attachments/assets/c9f6600f-6678-43ea-94c8-3dbd54d4351a)

And here is what it looks like on X11 with Cairo:

![image](https://github.com/user-attachments/assets/c06d3c00-d49f-46f3-8836-b15aeade44f0)

All of these 16x16 xpm images are "spilling over" and being drawn in a 17x17 area. This issue also does not happen on Windows, Mac, or X11 without Cairo.

I would love to figure out how to fix this because I think it looks quite ugly. Any ideas or suggestions are appreciated.

Thanks for reading.

**Comment by MatthiasWM** [comment, marked as answer]
As for the image overflow, I don't know much about Cairo, but I think the two following lines look suspicious:

Fl_Cairo_Graphics_Driver.cxx
```
802:    cairo_rectangle(cairo_, X - 0.5, Y - 0.5, W + 0.5, H + 0.5);
802:    cairo_rectangle(cairo_, X - 0.5, Y - 0.5, W, H);
814:    cairo_pattern_set_extend(pat, CAIRO_EXTEND_PAD);
814:    cairo_pattern_set_extend(pat, CAIRO_EXTEND_NONE);
```

Anyone

**Reply by MatthiasWM** [reply]
As for the fonts, yes, calling `Fl::set_fonts(NULL)` reinitializes fonts from scratch every time and may conflict with some caches. I don;t know the code well enough to make changes here, but I guess simply caching the number of fonts and returning that, if the function is called agin, should be enough.

**Reply by dannye** [reply]
> As for the fonts, yes, calling `Fl::set_fonts(NULL)` reinitializes fonts from scratch every time

Hm, I'm curious why Cairo is unique in this regard.

The Windows and X11 drivers do not have this same "limitation" (calling them repeatedly is "safe"):

https://github.com/fltk/fltk/blob/d51046cdeaf1dceac91e70fe6be358502280a431/src/drivers/GDI/Fl_GDI_Graphics_Driver_font.cxx#L114-L123

https://github.com/fltk/fltk/blob/d51046cdeaf1dceac91e70fe6be358502280a431/src/drivers/Xlib/Fl_Xlib_Graphics_Driver_font_x.cxx#L297-L299

https://github.com/fltk/fltk/blob/d51046cdeaf1dceac91e70fe6be358502280a431/src/drivers/Xlib/Fl_Xlib_Graphics_Driver_font_xft.cxx#L311-L312

But it's really not difficult to avoid the limitation, so not really a big problem.

**Reply by MatthiasWM** [reply]
In 293 of the Xlib driver, it does exactly what I suggest. We will fix it in any case, I would just like to get the authors opinion first.

**Reply by dannye** [reply]
Oh okay, sorry, I was slightly misinterpreting your previous comment as saying that the user application doing the caching "should be enough", that's my bad.

**Reply by dannye** [reply]
Regarding your suggested fixes for the xpm smearing, this change is enough to fix the problem completely for me:

```diff
-    cairo_pattern_set_extend(pat, CAIRO_EXTEND_PAD);
+    cairo_pattern_set_extend(pat, CAIRO_EXTEND_NONE);
```

Thank you very much for the suggestion! Looks great.

**Comment by MatthiasWM** [comment]
Never mind. I patched both bugs and they are hopefully gone now ;-)

**Reply by dannye** [reply]
Thank you! It's greatly appreciated :-)

**Comment by wcout** [comment]
Regression-Alert 🥴
Parts of these changes break things discussed in #537.
Have a look at the `mandelbrot`  demo in fractional scales:

![grafik](https://github.com/user-attachments/assets/5d59e082-3474-4a90-a29a-3680fd3c32ba)

**Reply by MatthiasWM** [reply]
Thanks. I will look at that asap

**Reply by MatthiasWM** [reply]
Ok, I see where the problem is. Reversing to CAIRO_EXTEND_PAD is good because Cairo uses pixels outside of actual image data to scale the image. The change in clipping alone already fixes @dannye 's issue of smearing pixmaps.

The clipping fix does make a problem in Mandelbrot visible. Scaling to a fractional value with Cairo generates edges with an alpha value, and rendering images data with an alpha value will alway let the pixels below "shine through". I will try to find a better clipping value.

**Reply by ManoloFLTK** [reply]
Here is my solution:
1) revert the changes in the clip rectangle and in the call to `cairo_pattern_set_extend()`
2) replace
  `if (need_extend) {`
by
  `if (need_extend && img->count() <= 2) { // exclude Fl_Pixmap's`

**Comment by MatthiasWM** [comment]
Please check the last commit. I reverted the image padding, which fixes the dim Mandelbrot image, and I also fixed the clipping rectangle by switching off antialiasing, which seems to fix the remaining rendering artifacts.

Does this fix everything, or did I miss a case?

**Reply by ManoloFLTK** [reply]
The last commit is OK with all my tests.

**Reply by dannye** [reply]
This seems to work just as well for me too :+1: Thanks
